### ISSUE TITLE
[HOLD for payment 2023-09-06] [$1000] Pasting on Web isn't automatically focusing on composer

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#25550]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a chat in newDot Web/Desktop
2. Remove focus from composer
3. Copy some text into your clipboard (`CMD+C`/`CTRL+C`)
4. With the chat open and the composer unfocused, paste the text you have in your clipboard (`CMD+V`/`CTRL+V`)
## Expected Result: 
The composer should be focused and the text should be pasted into it.
## Actual Result:
 Nothing happens

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.55-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/36201da6-6677-4592-b5d3-6166dd7f9a36

**Expensify/Expensify Issue URL:**
**Issue reported by:** @PauloGasparSv
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1691623314942259

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ed32306975a2bc2a</li>
        <li>Upwork Job ID: 1693708827153092608</li>
        <li>Last Price Increase: 2023-08-21</li>
<li>Automatic offers: </li>
<ul>
<li>s77rt | Reviewer | 26293575</li>
<li>joh42 | Contributor | 26293576</li>
</ul></li>
    </ul>
</details>

**Comment by b4s36t4** [comment]
I think this intended, we've stopped it. https://github.com/Expensify/App/issues/25409#issuecomment-1682764840

**Comment by NicMendonca** [comment]
@b4s36t4 the feature was suppressed in this PR: https://github.com/Expensify/App/pull/23525

**Comment by b4s36t4** [comment]
@NicMendonca So do we want to enable it now?

**Comment by s77rt** [comment]
@rayane-djouah Thanks for the proposal. Your RCA is not exactly right, we do have a global listener for paste events even if the composer is not focused (as long as its screen is focused). However we just return early and do nothing in this case..

Please note that the use of `setTimeout` is discouraged as it's mostly used as a workaround. I think your use case is a workaround here as we are waiting for an _undefined_ amount of time - We are just pushing the focus call to the stack.

**Comment by s77rt** [comment]
@tienifr Thanks for the proposal. Your RCA is correct. However your proposed solution is outdated. Please update it to reflect the current `main` - We should also support pasting files when pressing CTRL/CMD+V.

**Comment by s77rt** [comment]
@getusha Thanks for the proposal. Your RCA makes sense. However the solution does not seem to be working. I think the proposal is outdated. Please update it and tag me again..

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Pasting on Web isn't automatically focusing on composer
### What is the root cause of that problem?
We're manually blocking the paste events if the input is focused. Also we're blocking the any meta related events to trigger the focus here https://github.com/Expensify/App/blob/d09faa6b5704234a4e42a510cb2ce340c00ca515/src/pages/home/report/ReportActionCompose.js#L472

### What changes do you think we should make in order to solve the problem?

Here we're checking for https://github.com/Expensify/App/blob/d09faa6b5704234a4e42a510cb2ce340c00ca515/src/components/Composer/index.js#L268 for the composer focus.

We can update the function to focus if it the event is paste event which we can trigger by using the event from variable from here https://github.com/Expensify/App/blob/d09faa6b5704234a4e42a510cb2ce340c00ca515/src/components/Composer/index.js#L266-L268 and pass that event to the function here https://github.com/Expensify/App/blob/d09faa6b5704234a4e42a510cb2ce340c00ca515/src/pages/home/report/ReportActionCompose.js#L454.

### Changes

```jsx
const isVisible = checkComposerVisibility(event);
```

```jsx
// Focus the composer if it is not focused or it's not coming from paste event (e can be empty) && check for the paste event details
if (!textInputRef.current.isFocused() && !_.isEmpty(e) && (e.metaKey || e.ctrlKey) && e.key === 'v') {
    focus();
    return true;
}
```
> Also we should update the dependencies of `checkComposerVisibility` function.

This won't break any previous issues like
* Email pasted twice
* Pasting in edit composer get's pasted into main composer
* Able to copy paste a file.




### What alternative solutions did you explore? (Optional)
NA

## Result

https://github.com/Expensify/App/assets/59088937/4d275f85-6683-482e-9535-674e9129accb

**Comment by getusha** [comment]
@s77rt since the component was updated to function component, i updated the code snippet accordingly.

# Proposal

[Updated](https://github.com/Expensify/App/issues/25550?notification_referrer_id=NT_kwDOBHjiV7M3NDQ0MzY3MzkyOjc1MDMxMTI3#issuecomment-1685842100)

**Comment by s77rt** [comment]
@joh42 Thanks for the proposal. Your RCA makes sense and the solution too. However I feel that we may be missing something. In https://github.com/Expensify/App/pull/23135 we had the early return check and it was working. It seems that something else has been changed that broke this functionality. Any idea on that?

**Comment by s77rt** [comment]
@b4s36t4 Thanks for the proposal. I think it's pretty much the same as @getusha's proposal.

**Comment by s77rt** [comment]
@getusha Thanks for the update. Unfortunately the solution does not seem to be working for me. The composer gets focused but the text is not pasted. Can you please double check? (and test on `main`)

[Kooha-2023-08-22-22-30-35.webm](https://github.com/Expensify/App/assets/16493223/87608500-24b1-4733-8714-1ebee9c54df6)

**Comment by s77rt** [comment]
@tienifr Sorry, my bad here. Your proposal is not outdated, I just misread the function callback (`paste` vs `handlePaste`). Yet I'm afraid that your solution would introduce a regression, if you type in an edit message input and paste a message it will get pasted on the main composer instead.

[Kooha-2023-08-22-22-32-41.webm](https://github.com/Expensify/App/assets/16493223/563d5de5-b691-45d1-8d5e-46a9a4043189)

**Comment by getusha** [comment]
@s77rt on which OS are you testing this? and which browser?

**Comment by joh42** [comment]
@s77rt this is the breaking change, there was even a comment mentioning it:
https://github.com/Expensify/App/pull/23525/files#r1273979811

Previously, we would check if `this.textInput.id !== event.target.id`.

I reimplemented that check just to test it, and found that neither `this.textInput` nor the event target (usually) had an id. As such we would not return early, event though the textInput was not the target of the paste event.

**TLDR: That early return was actually incorrect, as the paste would go through if neither the textInput nor the event target had an id**

**Comment by joh42** [comment]
@s77rt Here is a video demonstrating what I mean (sorry for the horrible quality lol, compressed it so it was less than 10MB):

https://github.com/Expensify/App/assets/138504087/c9e4efa8-16f7-4389-8f18-d76b5e923a5a

**Comment by b4s36t4** [comment]
@s77rt yes it looks similar but the approach and the way is different means I'm passing an event only from Composer not from other places. 

Have you tried my solution?

**Comment by s77rt** [comment]
@getusha
> on which OS are you testing this?

I tested in Linux, it didn't work (even after using ctrlKey). But it works in macOS. However this does not work if we use the paste option from the menu bar.

**Comment by s77rt** [comment]
@joh42 That actually makes sense. Reverting that PR seems to help, the composer gets focused yet nothing is pasted - Probably unrelated. Overall the proposal looks good to me.

🎀 👀 🎀 C+ reviewed
[Link to proposal](https://github.com/Expensify/App/issues/25550#issuecomment-1687238351)

**Comment by s77rt** [comment]
@b4s36t4 I don't think the approach is different, we will still have to pass the event from `focusComposerOnKeyPress`.

> Have you tried my solution?

Yes, same notes as @getusha's proposal, it does not work if paste is clicked from the menu bar.

**Comment by tgolen** [comment]
The C+ approved proposal looks OK to me. There is a question in there about adding a prop for `mainComposer` and I don't think it's necessary to do that at this point. Let's just keep it simple for now.

**Comment by joh42** [comment]
Thanks! You can expect a PR later today.

**Comment by joh42** [comment]
The [PR](https://github.com/Expensify/App/pull/25900) is now ready for review @s77rt

**Comment by b4s36t4** [comment]
Raised a bug here about the regression this would be causing. https://expensify.slack.com/archives/C049HHMV9SM/p1692987504856949 @s77rt

**Comment by s77rt** [comment]
@b4s36t4 It does not seem to be a regression from this PR. I can reproduce even with the PR reverted.

https://github.com/Expensify/App/assets/16493223/4032e800-97b5-4ea6-93a2-fa260c945692

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: https://github.com/Expensify/App/pull/23525
- The offending PR has been commented on: https://github.com/Expensify/App/pull/23525#discussion_r1310277656
- A discussion in #expensify-bugs has been started: I don't think this is needed for this case.
- Determine if we should create a regression test for this bug: Yes

___

**Regression Test Proposal** (Web/Desktop Only)
1. Open App
2. Go to any chat
3. Copy any message
4. Click anywhere outside the composer
5. Verify the composer is not focused
6. Press CMD+V
7. Verify the composer is focused and text is pasted correctly

**Comment by NicMendonca** [comment]
- Reporter: N/A
- Contributor: @joh42 - $1,500
- Contributor +: @s77rt - $1,500

**Comment by NicMendonca** [comment]
@joh42 you've been paid! 🎉

**Comment by NicMendonca** [comment]
> Reporter: N/A
> Contributor: @joh42 - $1,500
> Contributor +: @s77rt - $1,500 (via Upwork)

**Comment by NicMendonca** [comment]
all set here ✅
